### PR TITLE
Added recipe for plain function

### PR DESCRIPTION
**Pull request by jarrettmeyer**
This fixes issue #481. Adds documentation recipe for functions vs. beforeEach.

### CODE REVIEW

**Comment by vdemedes** [comment]
I wouldn't focus on "stylistic" side of manually doing setup/teardown. I'd suggest focusing on actual benefits it gives to developers, which you actually mentioned at the end:

> 1. Setup functions allow for easier modifications between test runs.

If the recipe focuses on a code style, it's not really a recipe, it's just an opinion/preference. Recipes are meant to be official, ready-to-go solutions for questions/problems, that often occur while using AVA.

Everything else looks good, thanks @jarrettmeyer, looking forward to checking out and merging your updated recipe!

**Comment by jarrettmeyer** [comment]
@vdemedes Your recommendation is a change to the supporting text? Are the examples good/bad/necessary?

**Comment by vdemedes** [comment]
Yes, the supporting text. I think the code examples are fine. But I'd focus reader's attention to the point, that using manual setup/teardown functions gives you more flexibility. AVA's before/after hooks can't be turned off for certain tests, so here's exactly where "manual work" pays off.

The point of style, looking better and having straightforward tests is secondary, but also worth mentioning.

**Comment by vdemedes** [comment]
And @jamestalmage pointed out another advantage:

> The real advantage of the setup function approach is that you could have more than one, and call them based on what you need in a particular test.

File: docs/recipes/setup-vs-beforeEach.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
+# Setup Functions vs. beforeEach Functions
```
Comment (sindresorhus) [inline docs/recipes/setup-vs-beforeEach.md]:
```
# Setup functions vs. `beforeEach` hooks
```

**Comment by sindresorhus** [comment]
I would also at the end mention some of the benefits of using `beforeEach`. From #481:

> Still, there are benefits of beforeEach too. They're not mutually exclusive. beforeEach has benefits like supporting non-promise asynchronicity. Better output when a hook fails. Void hook usage doesn't require a call in each test. And more.

There might be more I haven't thought of.

File: docs/recipes/setup-vs-beforeEach.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+  t.true(await isValid);
+});
+
+test('authenticating with an invalid username', async t => {
+  t.context.credentials.username = 'bad_username';
+  const isValid = t.context.authenticator.authenticate(t.context.credentials);
+  t.false(await isValid);
+});
+
+test('authenticating with valid credentials', async t => {
```
Comment (forresst) [inline docs/recipes/setup-vs-beforeEach.md]: Replace _'authenticating with valid credentials'_ by _'authenticating with **in**valid credentials'_ ?

File: docs/recipes/setup-vs-beforeEach.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  t.true(await isValid);
+});
+
+test('authenticating with an invalid username', async t => {
+  const {authenticator, credentials} = setup({username: 'bad_username'});
+  const isValid = authenticator.authenticate(credentials);
+  t.false(await isValid);
+});
+
+test('authenticating with valid credentials', async t => {
```
Comment (forresst) [inline docs/recipes/setup-vs-beforeEach.md]: Replace 'authenticating with valid credentials' by 'authenticating with invalid credentials' ?

**Comment by sindresorhus** [comment]
@jarrettmeyer ping :)

**Comment by jamestalmage** [comment]
@jarrettmeyer Do you have time to follow up on the feedback? Would like to merge this soon.

**Comment by sindresorhus** [comment]
Closing for lack of response by OP.

**Pull request closed without merging**